docs: new tutorial "validate asyncapi documents using studio"

starlightknown opened the pull request:
<!--   Thank you for your contribution. Before you submit the pull request:
1. Follow our contribution guidelines
2. Test your changes and attach their results to the pull request.
3. Update the relevant documentation.
-->

**Description**

- NEW tutorial, "How to Validate AsyncAPI documents using studio"

**Related issue(s)**
<!-- If you refer to a particular issue, provide its number, othewise, remove this section.
For example, `Resolves #123`, `Fixes #43`, or `See also #33`. The `See also #33` option will not automatically close the issue after the PR merge. -->

#872

starlightknown committed:
Create validate-documents

starlightknown committed:
Update validate-documents

starlightknown commented:
@alequetzalli this is up for suggestions and reviews

starlightknown committed:
Update validate-documents

derberg commented on pages/docs/tutorials/validate-documents:
```diff
@@ -0,0 +1,144 @@
+---
```
you need to add `.md` to file name as now it do not show up in the preview

starlightknown committed:
added .md

changed name

starlightknown committed:
Merge branch 'master' into studio

quetzalliwrites committed:
Merge branch 'master' into studio

quetzalliwrites committed:
re-wrote content into tutorial writing format

There was a lot of unnecessary info that was not relevant to the core goal of the tutorial. The tutorial was also missing a sequential format from start to finish.

quetzalliwrites commented:
📝  Hey @starlightknown, a few feedback notes:  

1. This tutorial had a lot of unnecessary content that was directly copied over from the 'guide' PR. This should have been rewritten entirely and edited to be much more succinct. (Ex: mention of diff ways to validate, writing a description of an AsyncAPI document and its sections (we already have a Quickstart), and even talking about parsers)
2. This tutorial was not written in only sequential format, guiding us from steps 1-10. I re-wrote this content to show how tutorial steps should be written.


At this point, it would be great to have @derberg review it from a technical perspective and provide his feedback too on how to further improve this tutorial. I am curious to see if he thinks we should add more errors and what they might be. (I considered making an error in the `Payload` section (a number instead of the required string for email?) but wasn't 100% sure.

Thank you for all your work so far, Karuna! ✨✨✨✨

quetzalliwrites committed:
tiny fixes

quetzalliwrites commented:
Oh and we need add the Prev/Next steps buttons to this doc page too :)

starlightknown commented:
> Oh and we need add the Prev/Next steps buttons to this doc page too :)

would that lead to my next tutorial?

starlightknown committed:
Merge branch 'master' into studio

magicmatatjahu commented on pages/docs/tutorials/validate-documents.md:
```diff
@@ -0,0 +1,96 @@
+
+3. Open the console box that says `PROBLEMS`, so you can read all console errors.
+
+4. Fix the incorrect AsyncAPI specification number to `2.4.0` and the `version`number to `1.0.0`.
```
FYI Version `2.0.0` defined in `info.version` doesn't produce any error, because it can have any value. It should follow by [semver](https://semver.org/) format, but as I see we don't validate that 😆 I will create issue for it.

starlightknown replied:
Shall I keep this step in guide then? Or remove it

magicmatatjahu replied:
We can remove info about `version` field, yes :) Please also change `version` field to the `1.0.0` value in previous yamls, ok?

starlightknown replied:
done, thank you

magicmatatjahu commented on pages/docs/tutorials/validate-documents.md:
```diff
@@ -0,0 +1,96 @@
+  user/signedup:
+    subscribe:
+      message:
+      $ref: 'UserSignedUp'
```
YAML is based on indentations, so if you wanna define `$ref` in `message` you need to make indentation in this block, otherwise YAML understand it that `$ref` is defined in `subscribe` block, not `message`

```suggestion
        $ref: 'UserSignedUp'
```

magicmatatjahu commented on pages/docs/tutorials/validate-documents.md:
````diff
@@ -0,0 +1,96 @@
+
+4. Fix the incorrect AsyncAPI specification number to `2.4.0` and the `version`number to `1.0.0`.
+ 
+ ```
````
```suggestion
```yaml
```

magicmatatjahu commented on pages/docs/tutorials/validate-documents.md:
````diff
@@ -0,0 +1,96 @@
+4. Fix the incorrect AsyncAPI specification number to `2.4.0` and the `version`number to `1.0.0`.
+ 
+ ```
+ asyncapi: '2.4.0'
````
```suggestion
asyncapi: '2.4.0'
```

magicmatatjahu commented on pages/docs/tutorials/validate-documents.md:
````diff
@@ -0,0 +1,96 @@
+
+3. Copy and paste the below invalid AsyncAPI document:
+
+```
````
```suggestion
```yaml
```

magicmatatjahu commented on pages/docs/tutorials/validate-documents.md:
````diff
@@ -0,0 +1,96 @@
+
+6. Fix the `$ref` by changing it to: `'#/components/messages/UserSignedUp'`.
+
+```
````
```suggestion
```yaml
```

magicmatatjahu reviewed (commented):
I have few comments, just for your information :) And please use the right language in the code blocks because then the website renders it nicely with colors, without specifying the language you will see the unified color, which is not pleasant for the eyes :)

magicmatatjahu commented on pages/docs/tutorials/validate-documents.md:
```diff
@@ -0,0 +1,96 @@
+  user/signedup:
+    subscribe:
+      message:
+      $ref: 'UsersSignedUp'
```
```suggestion
        $ref: 'UsersSignedUp'
```

starlightknown committed:
Merge branch 'master' into studio

starlightknown committed:
Update pages/docs/tutorials/validate-documents.md

Co-authored-by: Maciej Urbańczyk <urbanczyk.maciej.95@gmail.com>

starlightknown committed:
Update pages/docs/tutorials/validate-documents.md

Co-authored-by: Maciej Urbańczyk <urbanczyk.maciej.95@gmail.com>

starlightknown committed:
Update pages/docs/tutorials/validate-documents.md

Co-authored-by: Maciej Urbańczyk <urbanczyk.maciej.95@gmail.com>

starlightknown committed:
Update pages/docs/tutorials/validate-documents.md

Co-authored-by: Maciej Urbańczyk <urbanczyk.maciej.95@gmail.com>

starlightknown committed:
Update pages/docs/tutorials/validate-documents.md

Co-authored-by: Maciej Urbańczyk <urbanczyk.maciej.95@gmail.com>

starlightknown committed:
Update pages/docs/tutorials/validate-documents.md

Co-authored-by: Maciej Urbańczyk <urbanczyk.maciej.95@gmail.com>

starlightknown committed:
Merge branch 'master' into studio

starlightknown committed:
removed version info

quetzalliwrites committed:
Merge branch 'master' into studio

quetzalliwrites commented on pages/docs/tutorials/validate-documents.md:
```diff
@@ -0,0 +1,96 @@
+This tutorial will teach you how to validate AsyncAPI documents using the [AsyncAPI Studio tool](https://studio.asyncapi.com/).
+
+We will start with a broken AsyncAPI document and troubleshoot via console errors step-by-step until we end up with a valid AsyncAPI document. This process will illustrate how to identify [`REQUIRED` properties in AsyncAPI documents](https://www.asyncapi.com/docs/reference/specification/v2.5.0#A2SObject).
+
```
@starlightknown Hey, you're missing the `Installation Guide` section. Please add it and tag us to let us know when it's ready to review again.

starlightknown replied:
Hey we discussed about this in slack and you told me that we don't need installation guide. This is done in studio so we don't need anything to install locally. Let me know your thoughts

quetzalliwrites replied:
Oooh you're right, we talked about this 😂. Sorry, my bad 🤪

quetzalliwrites commented:
> > Oh and we need add the Prev/Next steps buttons to this doc page too :)
> 
> would that lead to my next tutorial?

Yes @starlightknown, remember we're ordering the 4 new tutorials in the following order:
1. Create AsyncAPI document 
2. Validate AsyncAPI document 
3. Generate Code
4. Validate Events/Messages

starlightknown committed:
add: doc button

starlightknown commented:
> > > Oh and we need add the Prev/Next steps buttons to this doc page too :)
> > 
> > 
> > would that lead to my next tutorial?
> 
> Yes @starlightknown, remember we're ordering the 4 new tutorials in the following order:
> 
> 1. Create AsyncAPI document
> 2. Validate AsyncAPI document
> 3. Generate Code
> 4. Validate Events/Messages

added the buttons

![image](https://user-images.githubusercontent.com/74637789/199245666-f37b89a9-50b6-486a-b71d-50789d56915f.png)

akshatnema committed:
Merge branch 'master' into studio

akshatnema commented:
@starlightknown Is this PR still in progress for development or completed? Because I can still see that the routes you added in the Prev and UpNext buttons are missing. Are you working on those pages in this PR only?

starlightknown committed:
Merge branch 'master' into studio

starlightknown commented:
> @starlightknown Is this PR still in progress for development or completed? Because I can still see that the routes you added in the Prev and UpNext buttons are missing. Are you working on those pages in this PR only?

Hey, I added the routes too but it doesn't work maybe because we don't have other PR merged yet.

akshatnema commented:
> Hey, I added the routes too but it doesn't work maybe because we don't have other PR merged yet.

@starlightknown We actually can't merge this PR with undefined routes right now. If the other PRs are taking time to get merged, I'll suggest making the routes of the buttons according to the current scenario of the Docs tree. If you want to wait for other PRs, just we should make sure we don't merge this PR first. 

Also, please list down the PRs that are linked to this so that I could have a check on that also.

starlightknown commented:
> > Hey, I added the routes too but it doesn't work maybe because we don't have other PR merged yet.
> 
> @starlightknown We actually can't merge this PR with undefined routes right now. If the other PRs are taking time to get merged, I'll suggest making the routes of the buttons according to the current scenario of the Docs tree. If you want to wait for other PRs, just we should make sure we don't merge this PR first.
> 
> Also, please list down the PRs that are linked to this so that I could have a check on that also.

Those are #1025 and #1023. I have added the buttons according that. #1023 is create documents which is the previous tutorial, this is the current tutorial and the next is #1025

starlightknown committed:
add: next steps and buttons

starlightknown committed:
Merge branch 'master' into studio

derberg reviewed (commented):
I think in navigation it should show before code generation.

why?

flow should be
- write file
- validate file
- generate code

why, cause generator already has validator, so it makes no sense to validate after. But it makes sense to validate before, to not be surprised by validation errors from the generator.

---
Also how does this tutorial related to https://github.com/asyncapi/website/pull/971/files

---

Also I'm not a fan of showing validation only from Studio perspective. Many people prefer CLI, so instead of going to Studio they would prefer to do `asyncapi validate`  from the terminal - it is easier. Why not both? but CLI is more important imho

starlightknown commented:
So you recommend to add CLI methods in this tutorial?

derberg commented:
yes, definitely

but please align with @alequetzalli and @Annysah
because looks like now we will have the following tutorial flow:
- create AsyncAPI document
- validate document
- generate code
- validate messages

for `generate code` CLI is also needed. So I'm not 100% sure, but I think it doesn't make sense to have CLI installation instruction in both, `validate document` and `generate code`

quetzalliwrites committed:
Merge branch 'master' into studio

quetzalliwrites commented:
## Point 1: Prev/Next buttons 
@starlightknown, @akshatnema is 100% right here. Adding next/prev buttons without routes or routes that don't exist yet is not what you should do. We've talked about this offline and in our GSoD calls, but there are 2 ways you can handle this:
1. Include the routes of ONLY content currently LIVE and get your PR merged. Moving forward, as new content is merged, we will update the routes in those buttons. (not necessarily you, but whoever creates a PR that would require updating those buttons)
2. WAIT to merge this PR until ALL 4 NEW gsod tutorials are ready to merge. In this case, it would be ok to use routes for content that is not yet LIVE because you're waiting to merge it.

I'm glad that @akshatnema noticed this because we need to ensure each new doc page has working routes in the prev/next buttons. Please let us know if this still doesn't make sense, and I can totally hop on a call if that helps. 😄

## Point 2: Adding CLI commands to this tutorial
@derberg No, the purpose of this tutorial from the beginning has always been how to validate AsyncAPI documents with the Studio tool specifically. This does not mean we don't need a validation tutorial with CLI commands, it just means that is another topic. This tutorial is to showcase Studio validation, it should not go outside of the topic and discuss CLI commands. Each validation method requires its dedicated tutorial, focused and succinct to **one scenario**. 

At this point, it is too late to ask the GSoD project ALSO to add a 'CLI validation' tutorial, so you should [consider that work for me (Alejandra) that I pick up independently of GSoD via `/website PR #1094`](https://github.com/asyncapi/website/issues/1094). 😄✌🏽 

Let's keep this tutorial focused on Studio validation example and I will create a NEW Docs issue outside of GSoD scope for a NEW tutorial on AsyncAPI document validation via CLI commands. 

Something I will also point out is that creating a 2nd tutorial on the topic of validation methods opens the door for a new Tutorials sub-section on `validation` like so: `asyncapi.com/docs/tutorials/validation`. Personally, I have been looking forward to doing this for a while, and now that we have more content its a great time to make these improvements.


## Point 3: Not including CLI installation information in individual tutorials
@derberg I see you asked: `"So I'm not 100% sure, but I think it doesn't make sense to have CLI installation instruction in both, validate document and generate code" `

So to answer that question, YES, each individual tutorial will always include required installation items. It doesn't matter if one or more tutorials need the same items installed. It only matters that any tutorial can stand on it's own, and include any and all info needed to install for that tutorial. 

Never assume a user knows what to install or has already completed a previous tutorial with same installation items. 😜

quetzalliwrites committed:
Merge branch 'master' into studio

quetzalliwrites committed:
Merge branch 'master' into studio

derberg commented:
just keep in mind that if we split these, have separate Studio and separate CLI tutorial, we will repeat a lot of content, about 80%, intro, summary and background will be almost the same. AsyncAPI file the same. The only difference will in direct instructions on how to check stuff in Studio and how to do it with CLI, which is just `asyncapi validate <FILE_NAME>` 

and again, how is this PR related to https://github.com/asyncapi/website/pull/971 as I keep getting confused reviewing all these different PRs and sometimes seeing they overlapping each other 😅

quetzalliwrites commented:
> just keep in mind that if we split these, have separate Studio and separate CLI tutorial, we will repeat a lot of content, about 80%, intro, summary and background will be almost the same. AsyncAPI file the same. The only difference will in direct instructions on how to check stuff in Studio and how to do it with CLI, which is just `asyncapi validate <FILE_NAME>`
> 

That is fine, and having a short CLI tutorial is fine. Again, it's about keeping the separation of scenarios (tools).

so as an example, users will now be able to read them in an order like so...

`asyncapi.com/docs/tutorials/getting-started` **(our Quickstart)**
|
`asyncapi.com/docs/tutorials/create-asyncapi-document`
|
`asyncapi.com/docs/tutorials/validate-asyncapi-document/` **(shorter /directory name alternative: `/docs/tutorials/validate-document/`)**
|
`asyncapi.com/docs/tutorials/validate-asyncapi-document/studio-validation`
|
`asyncapi.com/docs/tutorials/validate-asyncapi-document/cli-validation`
|
`asyncapi.com/docs/tutorials/generate-code`
|
`asyncapi.com/docs/tutorials/validate-messages`

As a user, I can now see clearly that there is at least 2 tools/diff ways to validate my AsyncAPI documents. This is better than having to open 1 tutorial and dig around and find the diff ways I can do it. Then it wouldn't be a tutorial because it is talking about many diff ways to do something in diff scenarios.


> and again, how is this PR related to #971 as I keep getting confused reviewing all these different PRs and sometimes seeing they overlapping each other 😅

hahaha yeah they got confused and opened an extra PR.. not a big deal but I can see why you got confused 😄

quetzalliwrites committed:
Merge branch 'master' into studio

quetzalliwrites commented on pages/docs/tutorials/validate-documents.md:
```diff
@@ -0,0 +1,115 @@
+  suggestions={[
+    {
+      href: '/docs/tutorials/generate-code',
+      title: 'Generate Code',
```
```suggestion
      title: 'Generate code',
```

quetzalliwrites commented on pages/docs/tutorials/validate-documents.md:
```diff
@@ -0,0 +1,115 @@
+    },
+    {
+      href: '/docs/tutorials/validate-events',
+      title: 'Messages/events validation',
```
```suggestion
      href: '/docs/tutorials/message-validation',
      title: 'Message validation',
```

Heyo, can we [please update our file names](https://github.com/asyncapi/website/pull/1061/files) to this? we shouldn't use "events" since we already decided to go with the word `messages`. 😄✌🏽

starlightknown committed:
Update pages/docs/tutorials/validate-documents.md

Co-authored-by: Alejandra Quetzalli  <alejandra.olvera.novack@gmail.com>

starlightknown committed:
Update pages/docs/tutorials/validate-documents.md

Co-authored-by: Alejandra Quetzalli  <alejandra.olvera.novack@gmail.com>

starlightknown committed:
Update validate-documents.md

starlightknown committed:
Merge branch 'master' into studio

starlightknown commented:
All done!

derberg commented on pages/docs/tutorials/validate-documents.md:
```diff
@@ -0,0 +1,115 @@
+---
+title: "Studio AsyncAPI document validation"
```
should not titles follow some single rule, like call to action or something else

wouldn't it sound better as `Validate AsyncAPI document with Studio`?

starlightknown replied:
We thought of a sentence but it doesn't look good in the side navigation bar. We'll be getting a long sentence in two parts

derberg replied:
ok, but on my screen `Studio AsyncAPI document validation` is also split

<img width="315" alt="Screenshot 2022-11-22 at 14 46 25" src="https://user-images.githubusercontent.com/6995927/203329752-71fdff47-920a-4164-abc0-e544a849a280.png">

maybe we should consider making left nav wider. I kinda do not like that we need to adjust titles to fit screen, you know, it is like writing content for the browser not for the human 😄

starlightknown replied:
okay then I'll make it as Validate AsyncAPI document with Studio

derberg commented on pages/docs/tutorials/validate-documents.md:
```diff
@@ -0,0 +1,115 @@
+---
+title: "Studio AsyncAPI document validation"
+description: This tutorial will teach us how to validate AsyncAPI documents using the AsyncAPI Studio tool.
```
```suggestion
description: This tutorial will teach you how to validate AsyncAPI documents using the AsyncAPI Studio tool.
```

derberg commented on pages/docs/tutorials/validate-documents.md:
```diff
@@ -0,0 +1,115 @@
+# Introduction
+This tutorial will teach you how to validate AsyncAPI documents using the [AsyncAPI Studio tool](https://studio.asyncapi.com/).
+
+We will start with a broken AsyncAPI document and troubleshoot via console errors step-by-step until we end up with a valid AsyncAPI document. This process will illustrate how to identify [`REQUIRED` properties in AsyncAPI documents](https://www.asyncapi.com/docs/reference/specification/v2.5.0#A2SObject).
```
```suggestion
You will start with a broken AsyncAPI document and troubleshoot via console errors step-by-step until we end up with a valid AsyncAPI document. This process will illustrate how to identify [`REQUIRED` properties in AsyncAPI documents](https://www.asyncapi.com/docs/reference/specification/v2.5.0#A2SObject).
```

derberg commented on pages/docs/tutorials/validate-documents.md:
````diff
@@ -0,0 +1,115 @@
+
+3. Copy and paste the below invalid AsyncAPI document:
+
+```yaml
````
example should be based on the same file as the other tutorials.

also I think we should be consistent with other tutorials.

so say: "yo, if you go through this tutorial without performing previous steps, then get an asyncapi file with `asyncapi new --example=tutorial.yml --no-tty`. Paste it in Studio, all works perfect, documentation is rendered. Not let's have some fun to see how errors with your file are displayed. Go and break this and that in your document. You see not docs are not rendered and in diagnostics you see some more details"


otherwise we do not have tut 1 -> tut 2 -> tut 3 flow

derberg replied:
also `asyncapi: '2.7.0'` is risky, we might reach it april next year 😄 

maybe do `asyncapi: '1.0.0'`

starlightknown replied:
I'm not sure where to paste `asyncapi new --example=tutorial.yml --no-tty` in studio? I don't see any terminal or anything like that which can process this command @derberg

starlightknown replied:
![image](https://user-images.githubusercontent.com/74637789/203547408-d3f5bc7c-2b44-470b-984a-0a71f5206f81.png)
I see this in studio too, how useful will that be?

derberg replied:
this is something people would have to generate on local first, and copy/paste

although, we could also enable the same tutorial.yml in the Studio, when one clicks on 
<img width="240" alt="Screenshot 2022-11-24 at 12 04 39" src="https://user-images.githubusercontent.com/6995927/203768894-15a4d961-fb06-402b-82cb-6d7b56407486.png">

🤔

magicmatatjahu replied:
Do you wanna have invalid AsyncAPI document as template in the Studio? 😄

starlightknown replied:
oh actually if that helps but that will be used only for this tutorial "educational purposes only" right?

magicmatatjahu replied:
It's not for me to decide, but we can create templates only for tutorials, it's not a problem.

derberg replied:
> It's not for me to decide

@magicmatatjahu kinda is as you are one of Studio maintainers 😆 

but yeah, that would be marked as Tutorial with description that it is for education with docs only

magicmatatjahu replied:
Ok, should I create PR or @starlightknown? I can help.

derberg replied:
just folks remember, in studio we need current yaml 😄 

Just flow of this tutorial must indicate that user selects valid template and then just breaks the document in studio to simulate errors

derberg commented on pages/docs/tutorials/validate-documents.md:
````diff
@@ -0,0 +1,115 @@
+            description: Email of the user
+ ``` 
+
+# Troubleshoot Studio console errors
````
what console errors you mean?

derberg replied:
studio just changed a bit, have a look

starlightknown replied:
oh it's diagnostics now, fixed

derberg commented on pages/docs/tutorials/validate-documents.md:
```diff
@@ -0,0 +1,115 @@
+# Troubleshoot Studio console errors
+Let's fix the errors one by one until we end up with a valid AsyncAPI document.
+
+1. In Studio, open the **visualizer** found on the left-side menu.
```
why do I need to open visualizer to troubleshoot errors?

starlightknown replied:
fixed it

derberg reviewed (changes requested):
left some comments

please also fix section's headings

starlightknown committed:
Update pages/docs/tutorials/validate-documents.md

Co-authored-by: Lukasz Gornicki <lpgornicki@gmail.com>

starlightknown committed:
Update pages/docs/tutorials/validate-documents.md

Co-authored-by: Lukasz Gornicki <lpgornicki@gmail.com>

starlightknown committed:
Update validate-documents.md

starlightknown commented:
Notes (just for me)
**Validation keywords**
- type( string, integer, boolean, number, object, array)
- constant values
- objects (properties, additional properties)
- arrays (items, additional items)
- numbers (multipleOf, minimum, maximum)
- string (minlength, maxlenght, email)
-

starlightknown committed:
Update validate-documents.md

starlightknown committed:
Merge branch 'master' into studio

starlightknown committed:
Update validate-documents.md

starlightknown committed:
Update validate-documents.md

derberg commented on pages/docs/tutorials/validate-documents.md:
````diff
@@ -0,0 +1,126 @@
+2. Copy and paste the below invalid AsyncAPI document:
+
+```yaml
+{`asyncapi: '1.0.0'
````
hmm, where did you take this from, it refers to ancient asyncapi 1.0.0 😄 
also please remove 
```
{`
```
and 
```
`}
```

derberg commented on pages/docs/tutorials/validate-documents.md:
```diff
@@ -0,0 +1,126 @@
+            sentAt:
+              type: string
+              format: date-time
+              description: Date and time when the message was sent.`}
```
```suggestion
              description: Date and time when the message was sent.
```

derberg commented on pages/docs/tutorials/validate-documents.md:
```diff
@@ -0,0 +1,126 @@
+This tutorial taught us how to validate an AsyncAPI document using the AsyncAPI Studio tool. We also learned to troubleshoot an invalid AsyncAPI document by following the error message directions in diagnostics. In doing so, we learned how to identify `REQUIRED` properties in all AsyncAPI documents.
+
+## Next steps
+Now that you've completed this tutorial, go ahead to learn how to [validate messages/events]() which you will be sending to your application.
```
next step do not rever to the same doc as the buttons below, which is strange

also look at https://deploy-preview-1022--asyncapi-website.netlify.app/docs/tutorials/validate-documents#now-that-youve-completed-this-tutorial-go-ahead-to-learn-how-to-validate-messagesevents-which-you-will-be-sending-to-your-application

strange stuff with section, and to be honest I don't  know why 🤔 
try maybe `---` must have an empty line before

starlightknown replied:
fixed

derberg commented:
@magicmatatjahu you know studio and new layout the best, please do the technical review if all steps are accurate and all is properly explained 🙏🏼

magicmatatjahu reviewed (commented):
Everything is good for me :)

starlightknown committed:
Update pages/docs/tutorials/validate-documents.md

Co-authored-by: Lukasz Gornicki <lpgornicki@gmail.com>

starlightknown committed:
Update validate-documents.md

starlightknown committed:
Update validate-documents.md

derberg reviewed (dismissed):
technically all good, we just need to incorporate of course the new fragment with CLI installation

@alequetzalli leaving this one to you

@Annysah well done 👏🏼

starlightknown commented:
> technically all good, we just need to incorporate of course the new fragment with CLI installation
> 
> @alequetzalli leaving this one to you
> 
> @Annysah well done 👏🏼

Hehe it's me tho 😝

derberg commented on pages/docs/tutorials/validate-documents.md:
```diff
@@ -0,0 +1,133 @@
+
+If you did not follow the previous tutorial and do not have an `asyncapi.yaml` file ready, then generate one using `asyncapi new --example=tutorial.yml --no-tty` command.
+
+</Remember>
```
btw, I saw this in at least 2 tutorials already. Might make sense to turn it into fragment too

quetzalliwrites replied:
let's create new fragments for items longer than just 1 sentence tho

derberg commented:
ups @starlightknown I was suppose to comment it here https://github.com/asyncapi/website/pull/1025#pullrequestreview-1201090665 😄 

for this PR:

- just left one suggestion
- all good from my side, leaving it to @alequetzalli and also @magicmatatjahu in case of template in studio (not sure if @starlightknown what to try to contribute or not

anyway, my approval is not needed 👍

quetzalliwrites committed:
Merge branch 'master' into studio

starlightknown commented:
@magicmatatjahu I would like to add the template, could you please let me know what and where to do that?

magicmatatjahu commented:
@starlightknown 

You should create (exactly copy and paste) your invalid AsyncAPI file to this location https://github.com/asyncapi/studio/tree/master/src/examples - please create `tutorials` folder and there paste that document. 

Next thing is to create new example type in this file https://github.com/asyncapi/studio/blob/master/src/examples/index.tsx as `tutorials`. You should know what I mean - if not, let me know :)

And at the end you should update that file https://github.com/asyncapi/studio/blob/master/src/components/Modals/NewFileModal.tsx to render the new category, `Tutorials` - again, if you will have problems, let me know :)

quetzalliwrites committed:
Merge branch 'master' into studio

starlightknown committed:
Update validate-documents.md

starlightknown commented:
> @starlightknown
> 
> You should create (exactly copy and paste) your invalid AsyncAPI file to this location https://github.com/asyncapi/studio/tree/master/src/examples - please create `tutorials` folder and there paste that document.
> 
> Next thing is to create new example type in this file https://github.com/asyncapi/studio/blob/master/src/examples/index.tsx as `tutorials`. You should know what I mean - if not, let me know :)
> 
> And at the end you should update that file https://github.com/asyncapi/studio/blob/master/src/components/Modals/NewFileModal.tsx to render the new category, `Tutorials` - again, if you will have problems, let me know :)

Done!

starlightknown committed:
Merge branch 'master' into studio

starlightknown commented:
@magicmatatjahu helped me and we did it! Thank you
![image](https://user-images.githubusercontent.com/74637789/207032073-f52b803b-9b29-4746-9f7a-019365caecc3.png)

starlightknown committed:
Merge branch 'master' into studio

magicmatatjahu commented:
@starlightknown @derberg Let me know when you wanna merge that PR, I will merge PR in the Studio before this one.

derberg commented:
just wanted to clarify, as I see `Invalid Example` there, do we really want to have invalid example in studio?

magicmatatjahu commented:
@derberg 

> just wanted to clarify, as I see Invalid Example there, do we really want to have invalid example in studio?

I asked about it a few days ago and we agreed that it would be described in the Studio as a non-working template for teaching purposes :) Or I don't understand your concern.

derberg commented:
I suggested that we have a valid example, and that the flow of tutorial says, "you choose valid example and then just break it to get errors"

https://github.com/asyncapi/website/pull/1022#discussion_r1037156926

derberg commented:
having an invalid example is weird to me. This is why we also have a valid `tutorial.yml` in CLI. But if I'm the only one thinking it is weird, let's go with invalid. Thoughts?

starlightknown commented:
> having an invalid example is weird to me. This is why we also have a valid `tutorial.yml` in CLI. But if I'm the only one thinking it is weird, let's go with invalid. Thoughts?

If there's already valid and we break it then the user already knows which parts they broke, it'll be a whole cycle again and they won't try to fix coz they know the valid file too. if they have to do this in real life then they might not know as they did not actually fix an invalid file? 

Idk just my thought

magicmatatjahu commented:
as @starlightknown wrote. The user will start breaking down the specifications which is correct and do the tutorial backwards - why a tutorial then?

quetzalliwrites committed:
Merge branch 'master' into studio

quetzalliwrites commented:
my own two cents...

1. **The flow of this tutorial was always planned to begin with a BROKEN document that the user must learn how to debug by troubleshooting via console errors**. The goal is to learn how to troubleshoot when your code (document) is broken and you only have console errors to help guide you along the way. (It is pointless to begin with a CORRECT file that the user can easily know what section they broke. That is not how you manage these situations in the wild; **the point of this tutorial is to replicate a real bug scenario in Studio and show users how to use the tool to debug documents**.) 
2. Originally...I'm not sure if I'd have wanted to recommend having the sample buggy (broken) document in Studio, but the nice thing about this idea is that we could add a comment within the template that links to the tutorial. That way, users who accidentally run into this while using the Studio tool can immediately see we have a tutorial for learning how to debug their documents. If we do it this way, I think it's actually a neat idea from Macief. So I guess I like it? 😄

<img width="300" alt="Screen Shot 2022-12-15 at 5 02 49 PM" src="https://user-images.githubusercontent.com/19964402/207998483-51052a56-5e9f-4920-a8da-002f8f404879.png">

quetzalliwrites commented:
> I suggested that we have a valid example, and that the flow of tutorial says, "you choose valid example and then just break it to get errors"
> 
> [#1022 (comment)](https://github.com/asyncapi/website/pull/1022#discussion_r1037156926)

ya @derberg  you keep trying to do that but I keep telling you that is missing the point of the tutorial 😄😂😜

quetzalliwrites committed:
Merge branch 'master' into studio

starlightknown commented:
> my own two cents...
> 
> 1. **The flow of this tutorial was always planned to begin with a BROKEN document that the user must learn how to debug by troubleshooting via console errors**. The goal is to learn how to troubleshoot when your code (document) is broken and you only have console errors to help guide you along the way. (It is pointless to begin with a CORRECT file that the user can easily know what section they broke. That is not how you manage these situations in the wild; **the point of this tutorial is to replicate a real bug scenario in Studio and show users how to use the tool to debug documents**.)
> 2. Originally...I'm not sure if I'd have wanted to recommend having the sample buggy (broken) document in Studio, but the nice thing about this idea is that we could add a comment within the template that links to the tutorial. That way, users who accidentally run into this while using the Studio tool can immediately see we have a tutorial for learning how to debug their documents. If we do it this way, I think it's actually a neat idea from Macief. So I guess I like it? 😄
> 
> <img alt="Screen Shot 2022-12-15 at 5 02 49 PM" width="300" src="https://user-images.githubusercontent.com/19964402/207998483-51052a56-5e9f-4920-a8da-002f8f404879.png">

Shall I go ahead and add that comment?

starlightknown committed:
Update validate-documents.md

derberg commented:
@alequetzalli 

![](https://media.giphy.com/media/fQoxwZBVWq5jhLXRty/giphy.gif)

quetzalliwrites committed:
Merge branch 'master' into studio

quetzalliwrites commented:
> Shall I go ahead and add that comment?

@starlightknown oui ✨✨✨✨✨

starlightknown commented:
> > Shall I go ahead and add that comment?
> 
> @starlightknown oui sparklessparklessparklessparklessparkles

added, I guess this is good now?

starlightknown committed:
Merge branch 'master' into studio

quetzalliwrites commented:
hey @starlightknown, where is that comment added to the Studio invalid template example? can you please show me link to that

starlightknown commented:
> hey @starlightknown, where is that comment added to the Studio invalid template example? can you please show me link to that

https://github.com/asyncapi/studio/pull/536 it's not merged yet

magicmatatjahu commented:
@starlightknown @alequetzalli Should I merge the PR in Studio first or I should wait for this PR?

quetzalliwrites committed:
Merge branch 'master' into studio

quetzalliwrites commented:
> @starlightknown @alequetzalli Should I merge the PR in Studio first or I should wait for this PR?

@magicmatatjahu I recommend we wait to merge https://github.com/asyncapi/studio/pull/536 at the same time as this tutorial PR; otherwise, it will confuse users since the tutorial won't be live yet.

magicmatatjahu commented:
@alequetzalli Mentioned PR is already merged 😅 You can merge this one with one problem.

quetzalliwrites committed:
Merge branch 'master' into studio

quetzalliwrites commented:
> @alequetzalli Mentioned PR is already merged 😅 You can merge this one with one problem.

LOL !! yeah I see fran got excited and jumped the gun  to merge...

So you say "You can merge this one with one problem." Sorry, what problem is that? 🧐

starlightknown committed:
Update validate-documents.md

quetzalliwrites committed:
Merge branch 'master' into studio

quetzalliwrites committed:
Merge branch 'master' into studio

akshatnema commented on pages/docs/tutorials/validate-documents.md:
```diff
@@ -0,0 +1,125 @@
+## Introduction
+This tutorial will teach you how to validate AsyncAPI documents using the [AsyncAPI Studio tool](https://studio.asyncapi.com/).
+
+You will start with a broken AsyncAPI document and troubleshoot via console errors step-by-step until we end up with a valid AsyncAPI document. This process will illustrate how to identify [`REQUIRED` properties in AsyncAPI documents](https://www.asyncapi.com/docs/reference/specification/v2.5.0#A2SObject).
```
```suggestion
You will start with a broken AsyncAPI document and troubleshoot via console errors step-by-step until we end up with a valid AsyncAPI document. This process will illustrate how to identify [`REQUIRED` properties in AsyncAPI documents](https://www.asyncapi.com/docs/reference/specification/latest#A2SObject).
```

@alequetzalli @derberg shouldn't the link for the specification be directed to the latest specification, instead of just pointing to v2.5.0?

quetzalliwrites replied:
yes! thank you for that idea

quetzalliwrites committed:
Update pages/docs/tutorials/validate-documents.md

Co-authored-by: Akshat Nema <76521428+akshatnema@users.noreply.github.com>

quetzalliwrites committed:
fixed file name

file name didn't include tutorial validation method

quetzalliwrites committed:
Merge branch 'master' into studio

quetzalliwrites committed:
Merge branch 'master' into studio

quetzalliwrites committed:
fixed next steps content and links

quetzalliwrites committed:
Merge branch 'master' into studio

quetzalliwrites committed:
Merge branch 'master' into studio

quetzalliwrites committed:
Merge branch 'master' into studio

quetzalliwrites commented:
/rtm
